core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/CompactEntityComponent.scala

```diff
@@ -393,23 +395,22 @@ class CompactEntityQuery(driverComponent: DriverComponent)
    *
    * `query execution plan (for select): index range scan on idx_entity_type_name.`
    */
-  def copyEntities(sourceWorkspaceId: UUID, destWorkspaceId: UUID, refs: Set[EntityPointer]): ReadWriteAction[Int] =
-    if (refs.isEmpty) {
-      DBIO.successful(0)
-    } else {
-      val typeNameClauses = generateTypeNameSql(refs)
-      val sql = concatSqlActions(
-        sql"""insert into ENTITY(name, entity_type, workspace_id, record_version, deleted, attributes)
+  def copyEntities(sourceWorkspaceId: UUID, destWorkspaceId: UUID, refs: Set[EntityPointer]): WriteAction[Int] = {
+    val baseSQL = concatSqlActions(
+      sql"""insert into ENTITY(name, entity_type, workspace_id, record_version, deleted, attributes)
              select name, entity_type, $destWorkspaceId, record_version, 0, attributes
              from ENTITY e
              where e.workspace_id = $sourceWorkspaceId
-             and deleted = 0
-             and ( """,
-        reduceSqlActionsWithDelim(typeNameClauses.toSeq, sql" or "),
-        sql""" );"""
-      )
-      sql.asUpdate
+             and deleted = 0"""
+    )
+    if (refs.nonEmpty) {
+      val typeNameClauses = generateTypeNameSql(refs)
+      val entityTypeNameTuples = reduceSqlActionsWithDelim(typeNameClauses.toSeq, sql" or ")
+      concatSqlActions(baseSQL, sql" and (", entityTypeNameTuples, sql")").as[Int].head
+    } else {
+      baseSQL.as[Int].head
     }
+  }
```

Comment:
I find it confusing that in this public method, when `refs` is empty, it copies everything instead of nothing. I think I'd rather see an explicit "copyAllEntities" method which is called by clones and leave the existing "copyEntities(...refs)" method as-is, where it will copy nothing if nothing is specified.

I bet you could make a _private_ method containing an implementation called by both "copyAllEntities" and "copyEntities" which copies everything if its argument is empty, but for the public methods I think we should be more careful.